sensor/kubernetes/listener/resources/selector.go

```diff
@@ -1,45 +1,153 @@
+}
+
+func (s selectorWrap) getSelector() labels.Selector {
+	return s.selector
+}
+
+func (s selectorWrap) getNumLabels() uint {
+	return s.numLabels
+}
+
+func (s selectorWrap) getMatchNil() bool {
+	return s.matchNil
 }
 
 // selectorDisjunction is the disjunction (logical or) of a list of selectors.
-type selectorDisjunction []selector
+type selectorDisjunction []labels.Selector
 
-func (d selectorDisjunction) Matches(labels2 labels.Labels) bool {
+func (d selectorDisjunction) Empty() bool {
 	for _, sel := range d {
-		if sel.Matches(labels2) {
+		if !sel.Empty() {
+			return false
+		}
+	}
+	return true
+}
+
+func (d selectorDisjunction) String() string {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Add(r ...labels.Requirement) labels.Selector {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Requirements() (requirements labels.Requirements, selectable bool) {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) DeepCopySelector() labels.Selector {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) RequiresExactMatch(label string) (value string, found bool) {
+	panic("unused function")
+}
+
+func (d selectorDisjunction) Matches(labels labels.Labels) bool {
+	for _, sel := range d {
+		if sel.Matches(labels) {
 			return true
 		}
 	}
 	return false
 }
 
-// or returns the logical or of the given selectors.
+// or returns the logical or of the given SelectorWrappers.
 func or(sels ...selector) selector {
-	return selectorDisjunction(sels)
-}
-
-// SelectorFromMap converts the given map to a selector. It correctly translates a `nil` map to a `nothing` collector,
-// as is, e.g., used for headless services.
-func SelectorFromMap(labelsMap map[string]string) labels.Selector {
-	if len(labelsMap) == 0 {
-		return labels.Nothing()
+	var selWrapper = selectorWrap{nil, math.MaxUint, false}
+	var selectors selectorDisjunction
+	for _, s := range sels {
+		if s.getMatchNil() {
+			selWrapper.matchNil = true
+		}
+		if selWrapper.numLabels > s.getNumLabels() && (s.getNumLabels() > 0 || s.getMatchNil()) {
+			selWrapper.numLabels = s.getNumLabels()
+		}
+		selectors = append(selectors, s.getSelector())
+	}
+	if selWrapper.numLabels == math.MaxUint {
+		selWrapper.numLabels = 0
 	}
-	return labels.SelectorFromSet(labels.Set(labelsMap))
+	selWrapper.selector = selectors
+	return selWrapper
 }
 
-// MatcherOrEverything converts the given map to a selector. If the map is `nil` or empty it will translate to `everything`
-// selector. This is needed in cases like Network Policies where an empty PodSelector matches everything in the namespace.
-func MatcherOrEverything(labelsMap map[string]string) labels.Selector {
-	if len(labelsMap) == 0 {
-		return labels.Everything()
+// CreateSelector returns a SelectorWrapper for the given map of labels; matchNil determines whether
+// an empty set of labels matches everything or nothing.
+func createSelector(labelsMap map[string]string, matchNil bool) selectorWrap {
```

Comment:
>I would create two separate constructors where the difference in behavior is included in their names.

Interesting because that's how is used to be, and I refactored it into one function since I didn't like the code duplication. I can go back to two constructors if that makes it clearer, but can I ask why it's preferable?